fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt

```diff
@@ -0,0 +1,371 @@
+import org.mozilla.fenix.theme.FirefoxTheme
+import org.mozilla.fenix.theme.Theme
+
+private val minTabStripItemWidth = 160.dp
+private val maxTabStripItemWidth = 280.dp
+private val tabStripIconSize = 24.dp
+private val spaceBetweenTabs = 4.dp
+private val tabStripStartPadding = 8.dp
+
+/**
+ * Top level composable for the tabs strip.
+ *
+ * @param onHome Whether or not the tabs strip is in the home screen.
+ * @param onAddTabClick Invoked when the add tab button is clicked.
+ * @param onLastTabClose Invoked when the last tab is closed.
+ * @param onSelectedTabClick Invoked when a tab is selected.
+ */
+@Composable
+fun TabsStrip(
+    onHome: Boolean = false,
+    onAddTabClick: () -> Unit,
+    onLastTabClose: () -> Unit = {},
+    onSelectedTabClick: () -> Unit = {},
+) {
+    val browserStore = components.core.store
+    val tabsUseCases = components.useCases.tabsUseCases
+    val appStore = components.appStore
+    val isPrivateMode by appStore.observeAsState(false) { it.mode.isPrivate }
+    val state by browserStore.observeAsState(TabsStripState.initial) {
+        it.toTabsStripState(onHome = onHome, isPrivateMode = isPrivateMode)
+    }
+
+    TabStripContent(
+        state = state,
+        onAddTabClick = onAddTabClick,
+        onCloseTabClick = {
+            if (state.tabs.size == 1) {
+                onLastTabClose()
+            }
+            tabsUseCases.removeTab(it)
+        },
+        onSelectedTabClick = {
+            tabsUseCases.selectTab(it)
+            onSelectedTabClick()
+        },
+    )
+}
+
+@Composable
+private fun TabStripContent(
+    state: TabsStripState,
+    onAddTabClick: () -> Unit,
+    onCloseTabClick: (id: String) -> Unit,
+    onSelectedTabClick: (id: String) -> Unit,
+) {
+    Row(
+        modifier = Modifier
+            .fillMaxSize()
+            .background(FirefoxTheme.colors.layer1)
+            .systemGestureExclusion(),
+        verticalAlignment = Alignment.CenterVertically,
+    ) {
+        TabsList(
+            state = state,
+            modifier = Modifier.weight(1f, fill = false),
+            onCloseTabClick = onCloseTabClick,
+            onSelectedTabClick = onSelectedTabClick,
+        )
+
+        IconButton(onClick = onAddTabClick) {
+            Icon(
+                painter = painterResource(R.drawable.mozac_ic_plus_24),
+                modifier = Modifier.size(20.dp),
+                tint = FirefoxTheme.colors.iconPrimary,
+                contentDescription = stringResource(R.string.add_tab),
+            )
+        }
+    }
+}
+
+@Composable
+@OptIn(ExperimentalFoundationApi::class)
+private fun TabsList(
+    state: TabsStripState,
+    modifier: Modifier = Modifier,
+    onCloseTabClick: (id: String) -> Unit,
+    onSelectedTabClick: (id: String) -> Unit,
+) {
+    BoxWithConstraints(modifier = modifier) {
+        val listState = rememberLazyListState()
+        val tabWidth = ((maxWidth - tabStripStartPadding) / state.tabs.size) - spaceBetweenTabs
```

Comment:
could we get a small comment here explaining the reasoning behind the math for future maintainers?